### ISSUE TITLE
In Vite Dev Mode, request.url.port is incorrect

### ISSUE THREAD

**Issue by patdx** [remix-run/remix#8657]
### Reproduction

https://github.com/patdx/remix-vite-request-url-issue

### System Info

```shell
System:
    OS: Linux 6.6 Fedora Linux 39 (Workstation Edition)
    CPU: (4) x64 Intel(R) Core(TM) i5-4690 CPU @ 3.50GHz
    Memory: 10.66 GB / 23.40 GB
    Container: Yes
    Shell: 5.9 - /usr/bin/zsh
  Binaries:
    Node: 20.9.0 - /run/user/1000/fnm_multishells/39998_1706767571052/bin/node
    Yarn: 1.22.19 - /run/user/1000/fnm_multishells/39998_1706767571052/bin/yarn
    npm: 10.4.0 - /run/user/1000/fnm_multishells/39998_1706767571052/bin/npm
    pnpm: 8.14.3 - /run/user/1000/fnm_multishells/39998_1706767571052/bin/pnpm
  Browsers:
    Chrome: 121.0.6167.139
  npmPackages:
    @remix-run/dev: ^2.5.1 => 2.5.1 
    @remix-run/node: ^2.5.1 => 2.5.1 
    @remix-run/react: ^2.5.1 => 2.5.1 
    @remix-run/serve: ^2.5.1 => 2.5.1 
    vite: ^5.0.0 => 5.0.12
```

### Used Package Manager

pnpm

### Expected Behavior

In Vite Dev Mode, in Remix Loaders and Actions, `new URL(request.url).port` should be the port of the dev server, which is port 5173 by default.

### Actual Behavior

The value of `new URL(request.url).port` is actually the port of the requesting web browser app.

I believe this is happening because the [Vite Dev Handler](https://github.com/remix-run/remix/blob/2f9b39fb8ff335ca36914769e9b84fe1cc197481/packages/remix-dev/vite/node/adapter.ts#L36) is coded to use the Origin header when it exists over the Host header. The Origin header indicates the [origin (scheme, hostname, and port) that caused the request](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin).

The Host and Origin will conveniently have the same port number for a standard Remix Web App. However, if you are connecting to a Remix server from a browser web app on a different port, both port numbers would be different.

I'm not sure that it is ever necessary to use the `Origin` to calculate the `request.url`. Similar to the `createRemixRequest()` function maybe it is better to only use the `host` heaader or possibly `x-forwarded-host`

**Comment by pcattori** [comment]
Good find! Definitely something we can fix. Just curious, is this causing any errors or problems in your app? Seems odd to depend on the port

**Comment by patdx** [comment]
@pcattori yes, it is causing an error for me. My Remix app API routes return absolute urls inside their responses referring to other resources on the server. The generated URLs are now incorrect in Vite dev mode.

So, I'm working on an application with two parts:

1. Remix-powered admin dashboard and API server (Port 5173 in Dev Mode)
2. Angular SPA for end users (Port 4200 in Dev Mode)

The Angular app gets its data from the Remix app's API routes.

The relevant parts of the Remix app looks like this:

```
app/
├─ routes/
│  ├─ api.locations.ts
public/
├─ images/
│  ├─ pizza-planet.jpg
```

The Angular SPA fetches data from the Remix app at `http://localhost:5173/api/locations`, and the Remix app returns the following json:

```json
[{
  "id": 1,
  "name": "Pizza Planet",
  "thumbnail": "http://localhost:5173/images/pizza-planet.jpg"
}]
```

The Remix API server gives back the absolute URL so the Angular app can just plop that url into an `<image src="..." />` and it will simply work regardless of where the Angular app is served from.

So internally the Remix app code looks like this:

```ts
// /app/routes/api.locations.ts
export function loader({ request }) {
  return json({
    id: 1,
    name: "Pizza Planet",
    thumbnail: new URL("/images/pizza-planet.jpg", request.url).toString(),
  });
}
```

This is working fine in Remix Prod mode, and Remix Old Dev Mode. But in Remix Vite Dev Mode, when the Angular app requests the locations, it returns thumbnail url incorrectly as:

```
http://localhost:4200/images/pizza-planet.jpg
```

With the `4200` coming from the `Origin` header of the HTTP Request. Port 4200 maps back to the Angular Dev Server, but the image is actually served by Remix, so this URL doesn't actually load the image, it just returns a 404 error.

I feel that recent JavaScript libraries are converging around having the `Request.url` match the `Host` header  URL This is the case in Cloudflare Workers/Pages Functions, [H3](https://github.com/unjs/h3/blob/e38d4446354aba7e28d0971737f31787f25c4cad/src/utils/request.ts#L168), and [Hono](https://github.com/honojs/node-server/blob/1d75d95766f0f3eb5e089fcfc1022e6f05b9f497/src/request.ts#L107C35-L107C53). While it would not be feasible to account for every case of upstream reverse proxy (which may lead to headers like `x-forward-host`), I think it would be good to support some basic use cases like above, which could be done by reading the `Host` header and ignoring the `Origin` header. Hopefully this explanation makes a little more sense?

**Comment by brophdawg11** [comment]
Thank you for opening this issue, and our apologies we haven't gotten around to it yet!

With the release of [React Router v7](https://remix.run/blog/react-router-v7) we are sun-setting continued development/maintenance on Remix v2. If you have not already [upgraded to React Router v7](https://reactrouter.com/upgrading/remix), we recommend you do so. We've tried to make the upgrade process as smooth as possible with our [Future Flags](https://remix.run/docs/en/main/start/future-flags). We are now in the process of cleaning up outdated issues and pull requests to improve the overall hygiene of our repositories.

We plan to continue to address 2 types of issues in Remix v2:

- Bugs that pose security concerns
- Bugs that prevent upgrading to React Router v7

If you believe this issue meets one of those criteria, please respond or [create a new issue](https://github.com/remix-run/remix/issues/new?template=bug_report.yml).

For all other issues, ongoing maintenance will be happening in React Router v7, so:

- If this is a bug, please reopen this issue [in that repo](https://github.com/remix-run/react-router/issues/new?template=bug_report.yml) with a new [minimal reproduction](https://reactrouter.com/new) against v7
- If this is a feature request, please open a new [Proposal Discussion](https://github.com/remix-run/react-router/discussions/new?category=proposals) in React Router, and if it gets enough community support it can be considered for implementation

If you have any questions you can always reach out on [Discord](https://rmx.as/discord). Thanks again for providing feedback and helping us make our framework even better!
